src/encoder/ffmpeg.cpp

```diff
@@ -13,195 +13,213 @@ extern "C" {
-  }
+    AVDictionary*& get_ptr() {
+      return m_opts;
+    }
 
-  AVDictionary*& get_ptr() {
-    return m_opts;
-  }
+    std::string to_string() {
+      char* buffer = nullptr;
+      if (av_dict_get_string(m_opts, &buffer, '=', ',') < 0) {
+        std::free(buffer);
+        return {};
+      }
 
-  std::string to_string() {
-    char* buffer = nullptr;
-    if (av_dict_get_string(m_opts, &buffer, '=', ',') < 0) {
-      std::free(buffer);
-      return {};
+      std::string opts{ buffer };
+      return opts;
     }
 
-    std::string opts{buffer};
-    return opts;
-  }
+  private:
+    AVDictionary* m_opts = nullptr;
+  };
 
-private:
-  AVDictionary* m_opts = nullptr;
-};
+  AVCodecContext* Codec::make_context(const size_t kbits, AVCodec* codec, Size frame_size, std::size_t fps) {
 
-}
+    AVCodecContext* context = avcodec_alloc_context3(codec);
+    assert(context);
 
-namespace shar::codecs::ffmpeg {
+    std::fill_n(reinterpret_cast<char*>(context), sizeof(AVCodecContext), 0);
+    context->bit_rate = static_cast<int>(kbits * 1024);
+    context->time_base.num = 1;
+    context->time_base.den = static_cast<int>(fps);
+    context->pix_fmt = AV_PIX_FMT_YUV420P;
+    context->width = static_cast<int>(frame_size.width());
+    context->height = static_cast<int>(frame_size.height());
+    context->max_pixels = context->width * context->height;
 
-static AVCodec* select_codec(Logger& logger, const ConfigPtr& config){
-  const std::string codec_name = config->get<std::string>("codec", "");
-  if (!codec_name.empty()) {
-    if (auto* codec = avcodec_find_encoder_by_name(codec_name.c_str())) {
-      logger.info("Using {} encoder from config", codec_name);
-      return codec;
-    }
+    std::size_t divisor = std::gcd(frame_size.width(), frame_size.height());
+    context->sample_aspect_ratio.num = static_cast<int>(frame_size.width() / divisor);
+    context->sample_aspect_ratio.den = static_cast<int>(frame_size.height() / divisor);
 
-    logger.warning("Encoder {} requested but not found", codec_name);
+    return context;
   }
 
-  static std::array<const char*, 5> codecs = {
-      "h264_nvenc",
-      "h264_amf",
-      "h264_qsv",
-      // TODO: implement
-      //"h264_vaapi",
-      //"h264_v4l2m2m",
-      "h264_videotoolbox",
-      "h264_omx"
-  };
+  void Codec::select_codec(const ConfigPtr& config, Options& opts, Size frame_size, std::size_t fps) {
+
+    const std::string codec_name = config->get<std::string>("codec", "");
+    const std::size_t kbits = config->get<std::size_t>("bitrate", 5000);
+    if (!codec_name.empty()) {
+      if (auto* codec = avcodec_find_encoder_by_name(codec_name.c_str())) {
+        m_logger.info("Using {} encoder from config", codec_name);
+        m_encoder = codec;
+        m_context = make_context(kbits, codec, frame_size, fps);
+        if (avcodec_open2(m_context, codec, &opts.get_ptr()) >= 0) {
+          m_logger.info("Using {} encoder", codec_name);
+          return;
+        }
```

Comment:
`m_context` is not released in `else` case.  Also, you have removed the warning.

Reply:
I think we should have RAII wrappers for AVContext and AVCodec to avoid such errors.